price-feeder/oracle/provider/mexc.go

```diff
@@ -0,0 +1,504 @@
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse ticker volume")
+	}
+
+	p.tickers[symbol] = types.TickerPrice{
+		Price:  price,
+		Volume: volume,
+	}
+}
+
+func (p *MexcProvider) setCandlePair(candleResp MexcCandleResponse) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	close, err := sdk.NewDecFromStr(strconv.FormatFloat(candleResp.Metadata.Close, 'f', 5, 64))
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse candle close")
+	}
+	volume, err := sdk.NewDecFromStr(strconv.FormatFloat(candleResp.Metadata.Volume, 'f', 5, 64))
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse candle volume")
+	}
+	candle := types.CandlePrice{
+		Price:  close,
+		Volume: volume,
+		// convert seconds -> milli
+		TimeStamp: SecondsToMilli(candleResp.Metadata.TimeStamp),
+	}
+
+	staleTime := PastUnixTime(providerCandlePeriod)
+	candleList := []types.CandlePrice{}
+	candleList = append(candleList, candle)
+
+	for _, c := range p.candles[candleResp.Symbol] {
+		if staleTime < c.TimeStamp {
+			candleList = append(candleList, c)
+		}
+	}
+
+	p.candles[candleResp.Symbol] = candleList
+}
+
+func (p *MexcProvider) handleWebSocketMsgs(ctx context.Context) {
+	reconnectTicker := time.NewTicker(defaultMaxConnectionTime)
+	defer reconnectTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewWSMessage):
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message.
+				p.logger.Err(err).Msg("mexc: could not read message")
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz)
+
+		case <-reconnectTicker.C:
+			if err := p.disconnect(); err != nil {
+				p.logger.Err(err).Msg("error closing websocket")
+			}
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("mexc: error reconnecting")
+				p.keepReconnecting()
+			}
+		}
+	}
+}
+
+func (p *MexcProvider) disconnect() error {
+	err := p.wsClient.Close()
+	if err != nil {
+		return types.ErrProviderConnection.Wrapf("error closing Mecx websocket %v", err)
+	}
+	return nil
+}
+
+// reconnect closes the last WS connection then create a new one and subscribe to
+// all subscribed pairs in the ticker and candle pairs. If no ping is received
+// within 1 minute, the connection will be disconnected. It is recommended to
+// send a ping for 10-20 seconds
+func (p *MexcProvider) reconnect() error {
+	p.logger.Debug().Msg("mexc: reconnecting websocket")
+	wsConn, resp, err := websocket.DefaultDialer.Dial(p.wsURL.String(), nil)
+	defer resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("mexc: error reconnect to mexc websocket: %w", err)
+	}
+	p.wsClient = wsConn
+
+	currencyPairs := p.subscribedPairsToSlice()
+
+	telemetry.IncrCounter(
+		1,
+		"websocket",
+		"reconnect",
+		"provider",
+		string(ProviderMexc),
+	)
```

Comment:
if we are using something similar in other providers, then let's also create a helper function for this telemetry